parsl/monitoring/db_local.py

```diff
@@ -0,0 +1,218 @@
+          Column('task_stdin', Text, nullable=True),
+          Column('task_stdout', Text, nullable=True),
+    )
+
+
+def create_task_resource_table(meta):
+    return Table(
+          'task_resources', meta,
+          Column('task_id', Text, sa.ForeignKey('task.task_id'), nullable=False),
+          # Column('timestamp', Text, nullable=False, primary_key=True),
+          Column('timestamp', Text, nullable=False),
+          Column('task_run_id', Text, sa.ForeignKey('workflows.task_run_id'), nullable=False),
+          Column('psutil_process_pid', Text, nullable=True),
+          Column('psutil_process_cpu_percent', Text, nullable=True),
+          Column('psutil_process_memory_percent', Text, nullable=True),
+          Column('psutil_process_children_count', Text, nullable=True),
+          Column('psutil_process_time_user', Text, nullable=True),
+          Column('psutil_process_time_system', Text, nullable=True),
+          Column('psutil_process_memory_virtual', Text, nullable=True),
+          Column('psutil_process_memory_resident', Text, nullable=True),
+          Column('psutil_process_disk_read', Text, nullable=True),
+          Column('psutil_process_disk_write', Text, nullable=True),
+          Column('psutil_process_status', Text, nullable=True),
+    )
+
+
+class DatabaseHandler(logging.Handler):
+    """ The handler that takes a log record and puts it into an SQL database. Needs to be a bit fast. """
+    def __init__(self, elink):
+        """ Set up the handler to link it to the database specified by elink. """
+        logging.Handler.__init__(self)
+        self.eng = sa.create_engine(elink)
+        self.meta = sa.MetaData()
+        self.meta.reflect(bind=self.eng)
+
+    def emit(self, record):
+        """ Take a task record and insert or update information in the SQL database. Creates the necessary tables if needed.
+        Allow multiple tries since originally this was erroneous and used concurrently. Squashes errors/exceptions.
+        This handler is added to a logger that is returned by get_db_logger when requested. """
+        self.eng.dispose()
+        trys = 3
+        info = {key: value for key, value in record.__dict__.items() if not key.startswith("__")}
+        for t in range(trys):
+            # Having what i think is an issue to reflect so try a couple times and don't complain if breaks
+            failed = False
+            try:
+                with self.eng.connect() as con:
+                    # formating values to convert from python or parsl to db standards
+                    info['task_fail_history'] = str(info.get('task_fail_history', None))
+                    info['timestamp'] = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(record.created))
+                    run_id = info['task_run_id']
+
+                    # if workflow or task has completed, update their entries with the time.
+                    if 'time_completed' in info.keys() and info['time_completed'] != 'None':
```

Comment:
should that be `'None`` or None, the python none-value? if the former, that's weird...

Reply:
It is weird. This is the database handler which receives the logs from an HTTP request. The result is that everything was converted to a string to be sent over HTTP thus this is the string ''None''. I'm sure there are better ways of sending these logs over HTTP but this was how I got it to work. Maybe for a v2 I could switch away from all strings?